pkg/controllers/management/authprovisioningv2/role.go

```diff
@@ -153,6 +158,78 @@ func (h *handler) OnChange(key string, rt *v3.RoleTemplate) (*v3.RoleTemplate, e
 	return rt, nil
 }
 
+func (h *handler) hasAnnotationsForDeletingCluster(annotations map[string]string, isClusterRole bool) (bool, error) {
+	clusterName, cOK := annotations[clusterNameLabel]
+	clusterNamespace, cnOK := annotations[clusterNamespaceLabel]
+	if !cOK || (!cnOK && !isClusterRole) {
+		// if this role doesn't have a clustername/namespace label it isn't one of the protected roles, so move on
+		return false, nil
+	}
+
+	mgmtCluster, mgmtErr := h.mgmtClusters.Get(clusterName)
+	if mgmtErr != nil && !apierror.IsNotFound(mgmtErr) {
+		return false, mgmtErr
+	}
+
+	cluster, err := h.clusters.Get(clusterNamespace, clusterName)
+	if err != nil && !apierror.IsNotFound(err) {
+		return false, err
+	}
+	if apierror.IsNotFound(mgmtErr) && apierror.IsNotFound(err) {
+		return false, nil
+	}
+	if cluster == nil && mgmtCluster == nil {
+		// cluster + err both being nil should not happen - but if it does, safe to delete the role
+		return false, nil
+	}
+	// if the err is nil, we can assume that the cluster is not nil
+	if (cluster != nil && cluster.DeletionTimestamp == nil) && (mgmtCluster != nil && mgmtCluster.DeletionTimestamp == nil) {
+		// if the cluster isn't being deleted, it wasn't the source of the delete, so we can remove the role
+		return false, nil
+	}
+	return true, nil
+}
+
+func (h *handler) OnRemoveRole(key string, role *rbacv1.Role) (*rbacv1.Role, error) {
+	shouldEnqueue, err := h.hasAnnotationsForDeletingCluster(role.Annotations, false)
+	if !shouldEnqueue {
+		return role, err
+	}
+	// Enqueue to ensure perms remain until full delete. Err skips don't re-enqueue, so we do that manually
+	h.roleController.EnqueueAfter(role.Namespace, role.Name, reenqueueTime)
+	return role, generic.ErrSkip
+}
+
+func (h *handler) OnRemoveRoleBinding(key string, roleBinding *rbacv1.RoleBinding) (*rbacv1.RoleBinding, error) {
+	shouldEnqueue, err := h.hasAnnotationsForDeletingCluster(roleBinding.Annotations, false)
+	if !shouldEnqueue {
+		return roleBinding, err
+	}
+	// Enqueue to ensure perms remain until full delete. Err skips don't re-enqueue, so we do that manually
+	h.roleBindingController.EnqueueAfter(roleBinding.Namespace, roleBinding.Name, reenqueueTime)
+	return roleBinding, generic.ErrSkip
```

Comment:
```suggestion
	shouldEnqueue, err := h.hasAnnotationsForDeletingCluster(roleBinding.Annotations, false)
         if err != nil {
             return role, err
          }
          if shouldEnqueue{
              // Enqueue to ensure perms remain until full delete. Err skips don't re-enqueue, so we do that manually
	     h.roleBindingController.EnqueueAfter(roleBinding.Namespace, roleBinding.Name, reenqueueTime)
          }
	
	return roleBinding, generic.ErrSkip
```

Reply:
When using the hasAnnnotation function, we should first check the error to verify the validity of the `shouldEnqueue` variable